Mathlib/Topology/Algebra/Module/ModuleTopology.lean

```diff
@@ -413,4 +413,84 @@ lemma _root_.ModuleTopology.eq_coinduced_of_surjective
 
 end surjection
 
+section Prod
+
+variable {R : Type*} [TopologicalSpace R] [Semiring R]
+variable {M : Type*} [AddCommMonoid M] [Module R M] [TopologicalSpace M] [hM : IsModuleTopology R M]
+variable {N : Type*} [AddCommMonoid N] [Module R N] [TopologicalSpace N] [hN : IsModuleTopology R N]
+
+/-- The product of the module topologies for two modules over a topological ring
+is the module topology. -/
+instance prod : IsModuleTopology R (M × N) := by
+  constructor
+  haveI : ContinuousAdd M := toContinuousAdd R M
+  haveI : ContinuousAdd N := toContinuousAdd R N
+  -- In this proof, `M × N` always denotes the product with its *product* topology.
+  -- Addition `(M × N)² → M × N` and scalar multiplication `R × (M × N) → M × N`
+  -- are continuous for the product topology (by results in the library), so the module topology
+  -- on `M × N` is finer than the product topology (as it's the Inf of such topologies).
+  -- It thus remains to show that the product topology is finer than the module topology.
+  refine le_antisymm ?_ <| sInf_le ⟨Prod.continuousSMul, Prod.continuousAdd⟩
+  -- Or equivalently, if `P` denotes `M × N` with the module topology,
+  let P := M × N
+  letI τP : TopologicalSpace P := moduleTopology R P
+  haveI : IsModuleTopology R P := ⟨rfl⟩
+  haveI : ContinuousAdd P := ModuleTopology.continuousAdd R P
+  -- and if `i` denotes the identity map from `M × N` to `P`
+  let i : M × N → P := id
+  -- then we need to show that `i` is continuous.
+  rw [← continuous_id_iff_le]
+  change @Continuous (M × N) P (_) τP i
+  -- But `i` can be written as (m, n) ↦ (m, 0) + (0, n)
+  -- or equivalently as i₁ ∘ pr₁ + i₂ ∘ pr₂, where prᵢ are the projections,
+  -- the iⱼ's are linear inclusions M → P and N → P, and the addition is P × P → P.
+  let i₁ : M →ₗ[R] P := LinearMap.inl R M N
+  let i₂ : N →ₗ[R] P := LinearMap.inr R M N
+  rw [show (i : M × N → P) =
+       (fun abcd ↦ abcd.1 + abcd.2 : P × P → P) ∘
+       (fun ab ↦ (i₁ ab.1,i₂ ab.2)) by
+       ext ⟨a, b⟩ <;> aesop]
+  -- and these maps are all continuous, hence `i` is too
+  fun_prop
+
+end Prod
+
+section Pi
+
+variable {R : Type*} [τR : TopologicalSpace R] [Semiring R]
+
```

Comment:
```suggestion
variable {R : Type*} [TopologicalSpace R] [Semiring R]
```

Reply:
Again this results from a clear-up of the proof and me not picking up the pieces. Thanks!

There is an argument for keeping named instances and that's that it's then possible to supply them later on with (`\tauR := moduleTopology`) or whatever); in the past I found this helpful but I'm getting better at doing these proofs now and they're looking more natural (for example the introduction of `P` defined to mean M x N but with a different topology so I need this crutch less.

Reply:
Yeah, I noted the history and I agree there are definitely times when it's useful to have a name for typeclass parameters.